### PR TITLE
IRIS-4231 | Sign out proxy and refactor Redirect URL logic

### PR DESCRIPTION
**Pull request by slayful**
## Links

* http://wikia-inc.atlassian.net/browse/IRIS-4231

### CODE REVIEW

File: server/app/facets/auth/signout.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import {getRedirectUrlForPost} from '../../lib/auth-url-factory';
+/**
+ * @param {Hapi.Request} request
+ * @param {*} reply
+ * @returns void
+ */
+export default function post(request, reply) {
+	// TODO IRIS-4012: Implement anti-CSRF token
+	if (!request.auth.isAuthenticated) {
+		reply('Cannot sign out an unauthenticated user.').code(HttpStatus.BAD_REQUEST);
```
Comment (kvas-damian) [inline server/app/facets/auth/signout.js]: I'd rather redirect user to final destination instead of displaying an error message.
i can imagine an use-case where user logouts in one tab and then does the same in the second tab. Following that scenario result in an error.

File: server/app/lib/redirect-validator.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * @param {string} whileListedDomain
+	 * @returns {boolean}
+	 */
+	return whiteListedDomains.some((whiteListedDomain) => {
+		return domain.indexOf(whiteListedDomain, domain.length - whiteListedDomain.length) !== -1;
+	});
+}
+
+export function hostIsWhitelistedOrMatchesCurrent(host, currentHost) {
```
Comment (kvas-damian) [inline server/app/lib/redirect-validator.js]: maybe `isWhitelistedHostOrMatchesCurrent`?

File: server/app/lib/auth-url-factory.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+function getPayloadRedirect(payload) {
+	if (payload && payload.redirect) {
+		return payload.redirect;
+	} else {
+		return null;
+	}
+}
+
+export function getRedirectUrlForPost(request) {
```
Comment (kvas-damian) [inline server/app/lib/auth-url-factory.js]: Can we unify `getValidRedirectUrl` and `getRedirectUrlForPost` and use only one method?

File: server/app/lib/url-utils.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * We only allow HTTP or HTTPS or no protocol. "javascript:" for example will be rejected.
+ * @param protocol
+ * @return {boolean}
+ */
+export function hasHttpOrNoProtocol(protocol) {
+	return hasHttpProtocol(protocol) || !protocol;
+}
+
+function isSubdomain(domain, host) {
```
Comment (kvas-damian) [inline server/app/lib/url-utils.js]: what do you think about renaming parameters `domain` and `subdomain`? `host` doesn't tell me anything here.

File: server/app/routes.js [hunk 5]
```diff
@@ -233,6 +234,12 @@ let routes,
 			}
 		},
 		{
+			method: 'POST',
+			// until VCL is complete /logout rather than /signout
```
Comment (kvas-damian) [inline server/app/routes.js]: Please add TODO and ticket number.

File: server/tests/unit/lib/url-utils.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	var testCases = [
+
+		{
+			host: 'wikia.com',
+			ssl: true,
+			expected: 'https://wikia.com',
+			description: ''
+		},
+		{
+			host: 'wikia.com',
```
Comment (kvas-damian) [inline server/tests/unit/lib/url-utils.js]: Can we add more testcases. For example testing subdomains?

File: server/tests/unit/lib/url-utils.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+QUnit.test('hasHttpProtocol', function (assert) {
+	var testCases = [
+		{
+			url: 'http:',
+			expected: true,
+			description: 'Regular HTTP protocol'
+		},
+		{
+			url: 'https:',
```
Comment (kvas-damian) [inline server/tests/unit/lib/url-utils.js]: can we add some test cases with real urls?

File: server/app/lib/auth-url-factory.js (hunk 3 above)
Comment (slayful) [inline server/app/lib/auth-url-factory.js]: Refactored it slightly so they share as little code as possible.

File: server/app/routes.js (hunk 5 above)
Comment (slayful) [inline server/app/routes.js]: I can replace it now.

File: server/tests/unit/lib/url-utils.js (hunk 7 above)
Comment (slayful) [inline server/tests/unit/lib/url-utils.js]: There are tests for whole URLs unit/../auth-url-factory.js

File: server/app/facets/auth/signout.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+		reply.redirect(redirectUrl).takeover();
+	} else {
+		signOutUser(request)
+			.then(result => {
+				// For redirect Hapi uses 302 status code which should correctly change the method to GET
+				reply.redirect(redirectUrl)
+					.state('access_token', result.token)
+					.takeover();
+			}).catch((result) => {
+				reply('Sign out unsuccessful').code(result.status);
```
Comment (kvas-damian) [inline server/app/facets/auth/signout.js]: shouldn't this message be translated? Personally I'd rephrase it to say what to do next. Currently user has no clue what to do.

File: server/tests/unit/lib/url-utils.js (hunk 7 above)
Comment (kvas-damian) [inline server/tests/unit/lib/url-utils.js]: ok

File: server/app/facets/auth/signout.js (hunk 8 above)
Comment (slayful) [inline server/app/facets/auth/signout.js]: Seems like it never had any proper handling (in MW). I'm creating a ticket to address that:
https://wikia-inc.atlassian.net/browse/IRIS-4248
